scripts/install-cni.sh

```diff
@@ -130,71 +130,190 @@ else
+}
+
+#######################################
+# Replaces `@ipv4Subnet', '@ipv6SubnetOptional` and `@ipv6RouteOptional` placeholders using `.spec.podCIDR` from node json and node's ipv6 from metadata server.
+# Arguments:
+#   node json from kube-apiserver
+#   node_ipv6_addr node's IPv6 address from GCE metadata server
+#######################################
+function fillSubnetsInCniSpecLegacyTemplate {
+  local node=$1
+  local node_ipv6_addr=$2
+
+  local primary_subnet
+  primary_subnet=$(jq -r '.spec.podCIDR' <<<"$node")
+
+  if is_ipv4_range "${primary_subnet:-}" ; then
+    echo "PodCIDR IPv4 detected: '${primary_subnet:-}'"
+    cni_spec=${cni_spec//@ipv4Subnet/[{\"subnet\": \"${primary_subnet:-}\"\}]}
+  elif is_ipv6_range "${primary_subnet:-}" ; then
+    echo "Primary IPv6 pod range detected '${primary_subnet:-}'. It will only work with new spec template."
+    exit 1
+  else
+    echo "Response from $node_url"
+    echo "$node"
+    echo "Failed to fetch PodCIDR from K8s API server, primary_subnet=${primary_subnet:-}. Exiting (1)..."
+    exit 1
+  fi
 
   if [ -n "${node_ipv6_addr:-}" ] && [ "${node_ipv6_addr}" != "null" ]; then
     echo "Found nic0 IPv6 address ${node_ipv6_addr:-}. Filling IPv6 subnet and route..."
+    POPULATE_IP6TABLES="true"
+    echo "ip6tables will be populated because IPv6 podCIDR is configured (from node interface)"
 
     cni_spec=${cni_spec//@ipv6SubnetOptional/, [{\"subnet\": \"${node_ipv6_addr:-}/112\"\}]}
     cni_spec=${cni_spec//@ipv6RouteOptional/, ${CNI_SPEC_IPV6_ROUTE:-{\"dst\": \"::/0\"\}}}
+  else
+    echo "No IPv6 address found for nic0. Clearing IPv6 subnet and route..."
+    cni_spec=${cni_spec//@ipv6SubnetOptional/}
+    cni_spec=${cni_spec//@ipv6RouteOptional/}
+  fi
+}
 
-    # Ensure the IPv6 firewall rules are as expected.
-    # These rules mirror the IPv4 rules installed by kubernetes/cluster/gce/gci/configure-helper.sh
+function fillSubnetsInCniSpec {
+  case "${CNI_SPEC_TEMPLATE_VERSION:-}" in
+    2*)
+      fillSubnetsInCniSpecV2Template "$1" "$2"
+      ;;
+    *)
+      fillSubnetsInCniSpecLegacyTemplate "$1" "$2"
+  esac
+}
 
-    if ip6tables -w -L INPUT | grep "Chain INPUT (policy DROP)" > /dev/null; then
-      echo "Add rules to accept all inbound TCP/UDP/ICMP/SCTP IPv6 packets"
-      ip6tables -A INPUT -w -p tcp -j ACCEPT
-      ip6tables -A INPUT -w -p udp -j ACCEPT
-      ip6tables -A INPUT -w -p icmpv6 -j ACCEPT
-      ip6tables -A INPUT -w -p sctp -j ACCEPT
-    fi
 
-    if ip6tables -w -L FORWARD | grep "Chain FORWARD (policy DROP)" > /dev/null; then
-      echo "Add rules to accept all forwarded TCP/UDP/ICMP/SCTP IPv6 packets"
-      ip6tables -A FORWARD -w -p tcp -j ACCEPT
-      ip6tables -A FORWARD -w -p udp -j ACCEPT
-      ip6tables -A FORWARD -w -p icmpv6 -j ACCEPT
-      ip6tables -A FORWARD -w -p sctp -j ACCEPT
-    fi
+STACK_TYPE=$(jq -r '.metadata.labels."cloud.google.com/gke-stack-type"' <<<"$response")
```

Comment:
this is the controversial part, please add a comment explaining the meaning of this label.

AFAIK since last discussion , this labels represents the cluster-stack-type as per https://kubernetes.io/docs/concepts/cluster-administration/networking/#cluster-network-ipfamilies

@sdmodi @basantsa1989 @jingyuanliang so we are all in the same place.

So the logging below as "node" stack type will be confusing, I think it is better if we use "Cluster stack type label"

Reply:
I thought this was to resolve migration, so the stack type of each node is frozen at node creation, not subject to change upon netd restarts / node restarts?

Reply:
i changed variable name and comment to clearly indicate that this is a CLUSTER_STACK_TYPE